### DISCUSSION TITLE
What apis are available in migrations?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3687] [Q&A] [answered]
I'd like to make a one-time setup migration something like below. It works as a hook, but not a migration (`$app` is not defined):

```js
// 1699198734_pockethost_defaults.js

migrate(
  (db) => {
    const form = new SettingsUpsertForm($app)
    const {
      PH_APP_NAME = 'Acme',
      PH_INSTANCE_URL = 'http://localhost:8090',
      PH_APEX_DOMAIN = 'pockethost.io',
    } = process.env
    form.meta.appName = PH_APP_NAME
    form.meta.appUrl = PH_INSTANCE_URL
    form.meta.senderName = PH_APP_NAME
    form.meta.senderAddress = `${PH_APP_NAME}@${PH_APEX_DOMAIN}`
    form.submit()
  },
  (db) => {
    // add down queries...
  },
)

```

**Comment by ganigeorgiev** [comment]
`$app` is currently not supported in the migrations (mainly for compatibility with the Go APIs and to avoid user mixing the migration transaction dao and `$app.Dao()`).

All goja migration bindings can be found in https://github.com/pocketbase/pocketbase/blob/4e51e393a2e3f8f683e077a3ab3477ffef3f1a9e/plugins/jsvm/jsvm.go#L156-L170 (_note that the definitions may change in the future as there are plans to reuse the vm in order to reduce the memory consumption in case of many JS migrations_).

For an example initializing default application settings you can check https://pocketbase.io/docs/js-migrations/#initialize-default-application-settings.

**Reply by benallfree** [reply]
> For an example initializing default application settings you can check https://pocketbase.io/docs/js-migrations/#initialize-default-application-settings.

🤦 sorry

**Reply by benallfree** [reply]
So, concerning timestamp names in migration files, I assume that defines the order in which migrations are executed. However, if the pocketbase init happens to be between other migrations that have already ran, what will happen? Will it still execute?

**Reply by ganigeorgiev** [reply]
Yes, unapplied migrations always execute.

_Note that the timestamps of system migrations will change in the future to have negative values to ensure that new system migrations are always executed before any custom user defined one, but most end users shouldn't be affected by this._

**Reply by benallfree** [reply]
Ok in that case, here's my migration. Do I need to do a version check or will this work with all PB versions?

```js
migrate(
  (db) => {
    const APP_DEFAULTS = {
      appName: 'Acme',
      appUrl: 'http://localhost:8090',
      senderName: 'Support',
      senderAddress: 'support@example.com',
    }

    const { PH_APP_NAME, PH_INSTANCE_URL, PH_APEX_DOMAIN } = process.env

    const dao = new Dao(db)

    const settings = dao.findSettings()
    if (!settings) {
      throw new Error(`Expected settings here`)
    }

    const fix = (field, newValue) => {
      if (!newValue || settings.meta[field] !== APP_DEFAULTS[field]) return
      settings.meta[field] = newValue
    }
    fix(`appName`, PH_APP_NAME)
    fix(`appUrl`, PH_INSTANCE_URL)
    fix(`senderName`, PH_APP_NAME)
    fix(`senderAddress`, `${PH_APP_NAME}@${PH_APEX_DOMAIN}`)

    dao.saveSettings(settings)
  },
  (db) => {
    // add down queries...
  },
)

```

(edited)

**Reply by ganigeorgiev** [reply]
It should work, but please note that there is no backward compatibility guarantee and as mentioned in previous discussions the currently exposed internals will be refactored in attempt to simplify them and address some of the users reported issues (eg. hook exec ordering, validations, api errors propagation, etc.).

**Reply by benallfree** [reply]
Would wrapping in a try/catch help? I don't want to brick the instance if these updates fail.

**Reply by benallfree** [reply]
Answering my own question - yes, this works even if APIs are unavailable:

```js
migrate(
  (db) => {
    try {
      const APP_DEFAULTS = {
        appName: 'Acme',
        appUrl: 'http://localhost:8090',
        senderName: 'Support',
        senderAddress: 'support@example.com',
      }

      $apis.staticDirectoryHandler()

      const { PH_APP_NAME, PH_INSTANCE_URL, PH_APEX_DOMAIN } = process.env

      const dao = new Dao(db)

      const settings = dao.findSettings()
      if (!settings) {
        throw new Error(`Expected settings here`)
      }

      const fix = (field, newValue) => {
        if (!newValue || settings.meta[field] !== APP_DEFAULTS[field]) return
        settings.meta[field] = newValue
      }
      fix(`appName`, PH_APP_NAME)
      fix(`appUrl`, PH_INSTANCE_URL)
      fix(`senderName`, PH_APP_NAME)
      fix(`senderAddress`, `${PH_APP_NAME}@${PH_APEX_DOMAIN}`)

      dao.saveSettings(settings)
    } catch (e) {
      console.error(`***error applying defaults: ${e}`)
    }
  },
  (db) => {
    // add down queries...
  },
)
```

**Reply by ganigeorgiev** [reply, marked as answer]
All unapplied migrations are executed in a transaction, meaning that if one fail the database will remain unchanged.

It is up to the developers to decide whether a migration is optional or required for the application to work.

If you want the app to ignore a migration error, then yes, adding `try/catch` in the `migrate` callbacks would avoid interrupting the `serve` execution.
